[Form] Add choice_translation_parameters option for ChoiceType form

VincentLanglet opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #36845 <!-- prefix each issue number with "Fix #", if any -->
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/13677
<!--
Replace this notice by a short README for your feature/bugfix. This will help people
understand your PR and can be used as a start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Never break backward compatibility (see https://symfony.com/bc).
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against branch master.
-->

VincentLanglet committed:
Add choice_translation_parameters option

VincentLanglet committed:
Add changelog

VincentLanglet committed:
Fix tests

xabbuh commented on src/Symfony/Bridge/Twig/Resources/views/Form/bootstrap_4_layout.html.twig:
```diff
@@ -201,6 +201,7 @@
         {%- for child in form %}
             {{- form_widget(child, {
                 parent_label_class: label_attr.class|default(''),
+                label_translation_parameters: choice_translation_parameters|default([]),
```
Shouldn't this be `label_translation_parameters: choice_translation_parameters|default(label_translation_parameters)` for backwards compatibility (same everywhere else)?

VincentLanglet replied:
@xabbuh I really don't know.

The fact that `label_translation_parameters` is used for the choices is documented nowhere. I feel like it's working this way just by luck and I don't know if it was intending to be like this.

`translation_domain` is used by default, but it's a generic option.
`label_translation_parameters` is not a generic option, it seems really specific to the `label`.

For example, `choice_attr` doesn't seems to use `label_attr` as default value...

xabbuh replied:
I am pretty sure that quite some people will have figured out that they could use the `label_translation_parameters` option to achieve this. I'd rather not break that even though it was not intended to be used like this.

VincentLanglet replied:
> I am pretty sure that quite some people will have figured out that they could use the `label_translation_parameters` option to achieve this.

This behaviour wasn't intend to work like this and was not documented. It could be considered as a bug.

> I'd rather not break that even though it was not intended to be used like this.

`attr_translation_parameters` doesn't fallback on `label_translation_parameters`
`help_translation_parameters` doesn't fallback on `label_translation_parameters`

This would be inconsistent that
`choice_translation_parameters` fallback on `label_translation_parameters`

Plus, it would introduce a bug on this new feature.

With something like
```
# translations/messages.en.yaml
    form.order.yes: 'I confirm my order to the company'
    form.order.no: 'I cancel my order to the company'
    form.order.label: 'I cancel my order to company'
```
```
$builder->add('id', null, [
        'label' => 'form.order.label'
        'label_translation_parameters' => [
            'company' => 'ACME Inc.',
        ],
        'choice' => [
            'form.order.yes' => true,
            'form.order.no' => false,
        ],
        'choice_translation_parameters' => [],
    ]);
```

I would expect to read, 
```
    form.order.yes: 'I confirm my order to the company'
    form.order.no: 'I cancel my order to the company'
    form.order.label: 'I cancel my order to ACME inc'
```
But I will get
```
    form.order.yes: 'I confirm my order to the ACME inc'
    form.order.no: 'I cancel my order to the ACME inc'
    form.order.label: 'I cancel my order to ACME inc'
```

VincentLanglet replied:
@xabbuh WDYT about

The fact that
- attr_translation_parameters doesn't fallback on label_translation_parameters
- help_translation_parameters doesn't fallback on label_translation_parameters

And
```
$builder->add('id', null, [
        'label' => 'form.order.label'
        'label_translation_parameters' => [
            'company' => 'ACME Inc.',
        ],
        'choice' => [
            'form.order.yes' => true,
            'form.order.no' => false,
        ],
        'choice_translation_parameters' => [],
    ]);
```
will be buggy

xabbuh replied:
What do you mean with that your last example will be buggy?

VincentLanglet replied:
With something like
```
# translations/messages.en.yaml
    form.order.yes: 'Yes, I confirm my order to the company'
    form.order.no: 'No, I cancel my order to the company'
    form.order.label: 'Do you confirm your order to company ?'
```
```
$builder->add('id', null, [
        'label' => 'form.order.label'
        'label_translation_parameters' => [
            'company' => 'ACME Inc.',
        ],
        'choice' => [
            'form.order.yes' => true,
            'form.order.no' => false,
        ],
        'choice_translation_parameters' => [],
    ]);
```

I would expect to read, 
```
    form.order.yes: 'Yes, I confirm my order to the company'
    form.order.no: 'No, I cancel my order to the company'
    form.order.label: 'Do you confirm your order to ACME inc ?'
```
But if I fallback to `label_translation_parameters`, I will get
```
    form.order.yes: 'Yes, I confirm my order to the ACME inc'
    form.order.no: 'No, I cancel my order to the ACME inc'
    form.order.label: 'Do you confirm your order to ACME inc ?'
```

VincentLanglet replied:
Friendly ping, 
Did you miss my answer @xabbuh ? Is my example clear ?

xabbuh replied:
I see what you mean, but I don't believe that something like that is very common as you will usually don't have placeholders like a plain `company` string, but `%company or `{company}` or the like instead.

And from my experience, people discover all these silently "working" things any there will be complaints if we break that now.

VincentLanglet replied:
`choice_attr` does not default on `label_attr`.
`choice_translation_domain` does not default on `translation_domain`.
`choice_translation_parameters` will default 'label_translation_parameters`.

Best way to lose consistency and lose developper when they use this option.

> And from my experience, people discover all these silently "working" things any there will be complaints if we break that now.

Working and break are opinionated words here.
![xkcd workflow](https://imgs.xkcd.com/comics/workflow.png)

Currently the fact that `label_translation_parameters` are used is not even consistent:
In the two following example, the parameters was currently `{}` which does mean that there is no use of `label_translation_parameters` for a select.
https://github.com/symfony/symfony/pull/36851/files#diff-ebe082b8f8e40b852f895d7712b30730R83
https://github.com/symfony/symfony/pull/36851/files#diff-ebe082b8f8e40b852f895d7712b30730R88

So If I want to avoid a opinionated BC-break, I just have to default to `label_translation_parameters` for half the case.

How are introduced BC-break in Symfony ?
Because this would need to be changed in Symfony 6.0 and master is the only available branch.

VincentLanglet replied:
Hi @xabbuh @nicolas-grekas

How does BC-break are introduce in symfony ?

Let's say I use `label_translation_parameters` as default value for `choice_translation_parameters` in order to be fully BC (even if the previous behavior wasn't consistent and wasn't documented).

I would like that the `choice_translation_parameters` doesn't default on `label_translation_parameters` in the next major. In order to be consistent with `attr_translation_parameters` and `help_translation_parameters`.

What should I do ? 
I can't make the PR on the 6.0 branch, since it doesn't exist.
Should I add a comment `// NEXT_MAJOR: REMOVE THE DEFAULT VALUE` ?

xabbuh replied:
The process would be to trigger a deprecation when the fallback becomes active. But I honestly do not have a good idea how to implement that as we do not know whether one of the fallbacks will actually be used.

VincentLanglet replied:
So I don't see better solution than a `NEXT_MAJOR` comment... I pushed a new commit.
Maybe creating an issue with the 6.0 milestone ?

xabbuh replied:
We should not silently change the behaviour in 6.0. So that doesn't look like a valid option to me.

VincentLanglet replied:
The only option I see would be throwing a deprecation if `choice_translation_parameters` is `[]` and `label_translation_parameters` is not `[]`. With a message like
> The `choice_translation_parameters` won't fallback on `label_translation_parameters` on Symfony 6.0. Please check that it will not affect you or add the `choice_translation_parameters` option to your form. 

The only defect is that you can't get rid of every deprecation if you use `label_translation_parameters` and you don't need the `choice_translation_parameters` option...

VincentLanglet commented:
Hi @xabbuh @nicolas-grekas 

How does BC-break are introduced in Symfony ?
See https://github.com/symfony/symfony/pull/36851#discussion_r442493071

I can't make the PR on the 6.0 branch, since it doesn't exist.
Should I add a comment // NEXT_MAJOR: REMOVE THE DEFAULT VALUE ?

VincentLanglet committed:
Avoid BC-break

VincentLanglet commented:
Just found a situation where a callback is needed if the translations parameters are depending on the choice.
I need to handle a callback then.

VincentLanglet commented:
Just found a situation where a callback is needed if the translations parameters are depending on the choice.
I need to handle a callback then.

@xabbuh Do you know how I can support something like
```
'choice_translation_parameters' => function ($choices, $key, $value) {
    return ['%foo%' => $choices->getFoo()];
}
```

VincentLanglet committed:
Make choice_translation_parameters similar to choice_attr and support callable

VincentLanglet commented:
Hi @xabbuh, I totally rework the PR and updated the https://github.com/symfony/symfony-docs/pull/13677

I change the way I implemented `choice_translation_parameters` to work the same way it works for `choice_attr`.
This allow to not pass the same translation_parameters to the choices.

WDYT about it ? :)

fabpot commented:
@VincentLanglet You should probably bump de dep of `symfony/form` to `^5.2` in Twig Bridge.

VincentLanglet committed:
Bump version

VincentLanglet commented:
> @VincentLanglet You should probably bump de dep of `symfony/form` to `^5.2` in Twig Bridge.

Indeed ! Thanks.

I think this PR is finally ready to review then. :)

fabpot reviewed (commented):
LGTM, @xabbuh Can you review this one?

xabbuh commented on src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php:
```diff
@@ -174,13 +174,14 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
+     * @param array|callable|Cache\ChoiceTranslationParameters $labelTranslationParameters The parameters used to translate the choice labels
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null, $labelTranslationParameters = [])
```
Adding the argument here is a BC break.

xabbuh commented on src/Symfony/Component/Form/ChoiceList/Factory/ChoiceListFactoryInterface.php:
```diff
@@ -76,12 +76,13 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, callable $va
      * @return ChoiceListView The choice list view
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null);
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null/*, $labelTranslationParameters = []*/);
```
We should trigger a deprecation for implementations of this interface that do not have the new argument.

fabpot replied:
@VincentLanglet It looks like you forgot to implement this feedback.

VincentLanglet replied:
@fabpot I added deprecation for all the three `createView` now. I hope it was the right way to do it.

xabbuh commented on src/Symfony/Component/Form/ChoiceList/Factory/DefaultChoiceListFactory.php:
```diff
@@ -69,7 +69,7 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, callable $va
      * {@inheritdoc}
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, callable $index = null, callable $groupBy = null, $attr = null, $labelTranslationParameters = [])
```
adding the argument is a BC break

xabbuh commented on src/Symfony/Component/Form/ChoiceList/Factory/PropertyAccessDecorator.php:
```diff
@@ -145,15 +145,16 @@ public function createListFromLoader(ChoiceLoaderInterface $loader, $value = nul
      * @return ChoiceListView The choice list view
      */
-    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null)
+    public function createView(ChoiceListInterface $list, $preferredChoices = null, $label = null, $index = null, $groupBy = null, $attr = null, $labelTranslationParameters = [])
```
adding the argument is a BC break

jvasseur commented:
Is this really needed now that we have https://github.com/symfony/symfony/pull/37670 ?

VincentLanglet committed:
Avoid BC break

VincentLanglet commented:
I removed the BC-break @xabbuh :)

VincentLanglet committed:
Add deprecation

VincentLanglet commented:
> Is this really needed now that we have #37670 ?

Even if the problem may be solved by your new feature, I think that adding consistency for the options are still needed.
If the translation_parameters exists for the label, or the help, it should exists for the choices too.

jvasseur commented:
> > Is this really needed now that we have #37670 ?
> 
> Even if the problem may be solved by your new feature, I think that adding consistency for the options are still needed.
> If the translation_parameters exists for the label, or the help, it should exists for the choices too.

I personally would be in favor of deprecated the other parameters and instead encourages the use of the `t` function instead of having to ways of doing the same thing. But that's just my personal opinion.

fabpot commented:
@jvasseur's idea looks interesting indeed.

VincentLanglet commented:
> @jvasseur's idea looks interesting indeed.

@fabpot Should I close this PR or will it still be possibly merged before considering deprecate using `label_translations_params` etc ?

fabpot commented:
I think we should add something that could be deprecated soon after. So, I propose to close this PR and I hope someone will work on a new PR for deprecations.

VincentLanglet commented:
> I think we should add something that could be deprecated soon after. So, I propose to close this PR and I hope someone will work on a new PR for deprecations.

I thought again about it and you may have close too quickly this PR @fabpot, can you re-open it ?

@jvasseur How Translatable can be used in this situation ?

Choices are supposed to be an array, https://symfony.com/doc/current/reference/forms/types/choice.html#choices
Translatable can't be used as key currently.

IMHO
- The feature currently doesn't exists
- We don't know how much time will it be needed before 1) introducing the feature with translatable, 2) deprecating the `label_translation_parameters`, `choice_translation_parameters`, ... options.

So we should considering merging this feature, because the issue https://github.com/symfony/symfony/issues/36845 still exists.

jvasseur commented:
@VincentLanglet you don't need to pass a translatable as key in the `choices` array, you need to use the https://symfony.com/doc/current/reference/forms/types/choice.html#choice-label option to transform what you put in the choice array into a translatable object. 

So yes, the feature does exists currently.

VincentLanglet commented:
> @VincentLanglet you don't need to pass a translatable as key in the `choices` array, you need to use the https://symfony.com/doc/current/reference/forms/types/choice.html#choice-label option to transform what you put in the choice array into a translatable object.
> 
> So yes, the feature does exists currently.

Oh I didn't think it that way.

But, I propose the feature
```
'choice' => [
    'form.order.yes' => true,
    'form.order.no' => false,
],
'choice_translation_parameters' => [
    'form.order.yes' => ['%company%' => 'ACME Inc.'],
    'form.order.no' => [],
],
```
And you recommend
```
'choice' => [
    'form.order.yes' => true,
    'form.order.no' => false,
],
'choice_label' => function($choices, $key, $value) {
    $parameters = [
        'form.order.yes' => ['%company%' => 'ACME Inc.'],
    ];

    return new Translatable($key, $parameters[$key] ?? []);
}
```
I have trouble to see this solution as simpler.
So It could take some time before the introduction of a deprecation for the first solution, if we had one some day.

I'm not sure that the "Translatable should be the only option" is the best strategy for the developer.

@xabbuh Do you have some time to review this PR ? :)

jvasseur commented:
Personally I tend to never use the keys in `choices` for labels and always define `choice_label` because I find it cleaner so I would do something like this :

```
'choices' => [
    true,
    false,
],
'choice_label' => function ($choices, $key, $value) {
    if ($value) {
        return t('form.order.yes', ['%company%' => 'ACME Inc.']);
    } else {
        return t('form.order.no'),
    };
}
```

Of course this example doesn't looks that good but in most cases your choices will contains the necessary data to construct the full translation.

And even if it is a bit more verbose I think it's a better solution since it keeps the translation key and the translation parameters close to each others.

VincentLanglet commented:
> Personally I tend to never use the keys in `choices` for labels and always define `choice_label` because I find it cleaner so I would do something like this :
> 
> ```
> 'choices' => [
>     true,
>     false,
> ],
> 'choice_label' => function ($choices, $key, $value) {
>     if ($value) {
>         return t('form.order.yes', ['%company%' => 'ACME Inc.']);
>     } else {
>         return t('form.order.no'),
>     };
> }
> ```
> 
> Of course this example doesn't looks that good but in most cases your choices will contains the necessary data to construct the full translation.
> 
> And even if it is a bit more verbose I think it's a better solution since it keeps the translation key and the translation parameters close to each others.

It's easier when it's a true/false case.

But if choice are let's say `[1, 2, 3, 4]`, you ends with
```
'choice_label' => function ($choices, $key, $value) {
    if ($value === 1) {
        return t('form.order.1', ['%company%' => 'ACME Inc.']);
    } elseif ($value === 2) {
        return t('form.order.2', ['%company%' => 'ACME Inc.']);
    } elseif ($value === 3) {
        return t('form.order.3', ['%company%' => 'ACME Inc.']);
    } else {
        return t('form.order.4'),
    };
}
```
It starts to be verbose.

And if it was the recommended way, Symfony wouldn't use the `key` of the `choices` as label.

I find your solution over-complicated, you don't like mine.
I don't see any issue to do the same thing different ways. It will satisfied more developer.

For instance `choice_label` can accept
```
function ($choice) {
   return $choice->getFoo();
}
```
And
```
'foo`
```
And none of these solution are deprecated.

fabpot commented:
@VincentLanglet Can you have a look at the failing tests?

VincentLanglet committed:
Fix test

VincentLanglet commented:
> @VincentLanglet Can you have a look at the failing tests?

I fixed the error I got with missing space in `resolved_form_type_1.txt`,
But maybe there should be something about the
```
trim_trailing_whitespace = true
```
line in the `.editorconfig` file. My IDE is just automatically removing the trailing whitespaces every time.

I also remove the deprecation I added since there is already one.
```
The "Symfony\Component\Form\ChoiceList\Factory\DefaultChoiceListFactory::createView()" method will require a new "array|callable $labelTranslationParameters" argument in the next major version of its interface "Symfony\Component\Form\ChoiceList\Factory\ChoiceListFactoryInterface", not defining it is deprecated.
```
But I don't know how to avoid getting this in tests. I'm getting this because the `DefaultChoiceListFactory::createView` doesn't have the `$labelTranslationParameters` in his definition, because it would be a BC-break.

@fabpot @xabbuh

fabpot commented:
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

VincentLanglet commented:
Sure https://github.com/symfony/symfony/pull/38469
